[TSC] Adding the Technical Release Process Document

prabhubalan opened the pull request:
@rabi-abdel  - Committing the release process document  (for the issue Fixes #1033) along with some diagrams. Please review.

pgoyal01 commented:
@prabhubalan Re: proposedmodel.png
1. If all of the work is performed on the Master (green) branch, then the "Cherry Pick" has to be from the Master to the next version of the Baldy release (with Hot Fixes) and the Tag would have changed to, say, "4.0.1 Baldy".
2. The Tag has major version, major version with significant changes and minor (hot fixes) (the 3 digits) but I think that in CNTT we do not expect releases "major version with significant changes" -- Correct?
3. Typo: The bottom right "Tag" should be "5.0.0. Hallasan"

prabhubalan commented:
> @prabhubalan Re: proposedmodel.png
> 
> 1. If all of the work is performed on the Master (green) branch, then the "Cherry Pick" has to be from the Master to the next version of the Baldy release (with Hot Fixes) and the Tag would have changed to, say, "4.0.1 Baldy".
@pgoyal01 - Thanks Pankaj. Updated it now
> 2. The Tag has major version, major version with significant changes and minor (hot fixes) (the 3 digits) but I think that in CNTT we do not expect releases "major version with significant changes" -- Correct?
@pgoyal01 - May not be required for Baldy. Let me request @rabi-abdel to add more comment on this. May be required for the future release, however, we will take a collective decision based on what we experience in CNTT. This will be evolving document.
> 3. Typo: The bottom right "Tag" should be "5.0.0. Hallasan"
@pgoyal01 - Thanks for pointing it out. Corrected now.

gkunz commented on doc/tech/release_notes/release_process.md:
```diff
@@ -1,23 +1,53 @@
 
-<a name="h2"></a>
-## H2
+To deliver a hot fix (also known as patch) into the latest release, simply apply the hot fix in the delivery branch and tag it (4.0.1-Baldy). Once the hot fix is delivered, cherry pick technique will be applied to carry over the changes into the current release. As shown in the below diagram.
```
as discussed on the called, I'd like to propose to add a sentence here or rephrase: the mind set should always be to fix a problem on master first - irrespective of existing release branches. So, by default a hotfix should go to master first. Only if it does not apply anymore, a "local" fix on a release branch is valid.

bfcohen replied:
Be aware that we are treating documentation as code, which is an imperfect analogy.

prabhubalan replied:
I remember in the weekly call that we discussed and concluded to deliver the hot fix (also called patch) into the latest available release (which is Baldy in this case). To perform the patch delivery, we simply perform the hot fix in the delivery branch, and tag it (4.0.1-Baldy) as per the community standard practice and guideline. I remember the same pattern is followed in ONAP as well. Once the hot fix is delivered (and everyone back to happiness), we will decide how to carry over the changes into the current release branch. Most probably cherry pick technique will apply (same shown in the figure). Only during the stabilization phase, all the fixes will go in "master" and cherry pick technique will apply for taking some of the fixes onto delivery branch. Hot fix process will be more the exception than the rule.

prabhubalan replied:
@gkunz - I remember in the weekly call that we discussed and concluded to deliver the hot fix (also called patch) into the latest available release (which is Baldy in this case). To perform the patch delivery, we simply perform the hot fix in the delivery branch, and tag it (4.0.1-Baldy) as per the community standard practice and guideline. I remember the same pattern is followed in ONAP as well. Once the hot fix is delivered (and everyone back to happiness), we will decide how to carry over the changes into the current release branch. Most probably cherry pick technique will apply (same shown in the figure). Only during the stabilization phase, all the fixes will go in "master" and cherry pick technique will apply for taking some of the fixes onto delivery branch. Hot fix process will be more the exception than the rule.

gkunz replied:
Hi @prabhubalan. I am referring to best practices established in mature upstream communities, such as OpenStack. According to the OpenStack documentation [1] on stable branches, all fixes must exist on master first "*unless the change simply does not make sense on master*" and are then backported to stable branches. Please see [2] (item 4) and [3].

I do see that we are handling documentation as code here - as pointed out by @bfcohen - so we can be a bit more relaxed about not "breaking" a stable release. However, some ways-of-working could still apply.

[1] https://docs.openstack.org/project-team-guide/stable-branches.html
[2] https://docs.openstack.org/project-team-guide/stable-branches.html#appropriate-fixes
[3] https://docs.openstack.org/project-team-guide/stable-branches.html#processes

kedmison replied:
I agree with @gkunz.  It makes sense to 
1. refer to established practices in a mature organization and 
1. patch master first, when possible, and cherry-pick from there.  Documentation fixes may wind up going to multiple active releases, and merge history may get confusing if there is a mesh of cherry-picking going on.
1. be pragmatic about the differences in code vs. documentation.

prabhubalan replied:
Hi @gkunz , @kedmison - I have updated them now. Merge / Approve if you are happy with the content.

pgoyal01 commented:
@prabhubalan Can we not use the terms "hot fixes" and "patches" which are associated more with computer code than documentation.  @bfcohen had made the same point earlier. If you notice, @gkunz is careful in using the term "fixes" and that may be the course to follow.

prabhubalan commented:
> @prabhubalan Can we not use the terms "hot fixes" and "patches" which are associated more with computer code than documentation. @bfcohen had made the same point earlier. If you notice, @gkunz is careful in using the term "fixes" and that may be the course to follow.

prabhubalan commented:
> > @prabhubalan Can we not use the terms "hot fixes" and "patches" which are associated more with computer code than documentation. @bfcohen had made the same point earlier. If you notice, @gkunz is careful in using the term "fixes" and that may be the course to follow.

@pgoyal01 - Thanks and corrected now.

rabiabdel commented:
@prabhubalan remove the label proposed from the diagram please so we get this merged. Thanks

prabhubalan commented:
> @prabhubalan remove the label proposed from the diagram please so we get this merged. Thanks

@rabiabdel - I have removed them now. Thanks!

prabhubalan commented:
> @prabhubalan thank you for addressing the comments!

@gkunz -Thanks Georg
